### DISCUSSION TITLE
Use of UnifiedJedis & JedisPooled

### DISCUSSION THREAD

**Discussion by dbeken85** [xetorthio/jedis#2868] [General]
Hello all,

I am repeating this question as a mail, so if it is a duplicate, please excuse me.

I want to use JSON as a Jedis request. I do not have multiple transactions, so I do not want to use pipeline or transaction.

I am reading that JedisPooled is a simplified version of JedisPool, and recommended, but I can find not a single worthy documentation on how to use it. JedisPooled and UnifiedJedis is needed for JSON.

I have a multithreaded scenario, where I do want to reuse a connection pool. I do not want to create a new connection for each request. 

The "old" way:
1- Initiate JedisPool once
2- Assume, we have a method which sends JSON Data:
Create a Jedis Object inside a try-with-resource block (but Jedis has no Autoclose, so how does this even work??), for every thread calling this method.
3- Call .close() on the jedis, once the message has been sent.

How do we do this with UnifiedJedis and JedisPooled
- have a single connection pool possible for all threads.
- don't reconnect for each connection.
- using json via UnifiedJson or JasonPooled.

I am getting frustrated by the lack of documentation and the lack of comments in the code.

Best Regards,
Dincer

**Comment by dbeken85** [comment]
So after looking into the code, I think this is the correct way.
```

/* Retrieve the Connection Pool. */
        PooledConnectionProvider pooledConnectionProvider = this.connPoolCache.getIfPresent(connection);
        if(pooledConnectionProvider == null) {
            LOGGER.info("No connection to destination '{}' exists. Creating a new connection pool.", connection);
            pooledConnectionProvider = new PooledConnectionProvider(new HostAndPort(connection, REDIS_PORT), this.jedisClientConfig);
        }
        JedisPooled jedisPooled = new JedisPooled(pooledConnectionProvider);
        jedisPooled.hset(key, teidMap);
        
```

We check if there exists a connection, if so we will use that Pool in the newly instantiated JedisPooled object. In the DefaultCommandExecutor, in the try-with-resource block, the connection is closed, i.e. returned back to the pool. 

```
try (Connection connection = provider.getConnection(commandObject.getArguments())) {
      return connection.executeCommand(commandObject);
    }

```
This would answer my question.

**Comment by dzoech** [comment]
Hey @dbeken85 unfortunately, I, too, have a hard time comprehending how JedisPooled is supposed to be used.

Do I have to initialize a connection pool as a singleton/static object and then pass this to the JedisPooled constructor?
What happens when I don't provide a constructor argument to JedisPooled? Does it create a new connection pool for every JedisPooled instance or is there some shared pool abstracted somewhere I don't even have to configure?

**Comment by vedev9** [comment]
I'm also facing this doubt.

I did like this:

1) In MyDB class i declared a JedisPooled variable (but UnifiedJedis seems to apply as well). I also defined a method to initialize it.

```
public MyBD() {
    public static JedisPooled client;

    public void init(ConnectionPoolConfig someConf){
        client = new JedisPooled(someConf, dbUrl, dbPort);
    }
}
```


2) in my App class, in the main function i instantiate a MyDB object and call init on it, then inject the object everywhere is needed (e.g. in my handlers, since my app is a API server).

3) every time a handler is called (in a new thread) it has access to the singleton JedisPooled object and can perform the jsonGet and jsonSet i need.


But the doubts remain, is it the right approach for my usecase? Does the JedisPooled get a connection from the pool every time its methods are called and then releases it when done automatically? 
Any hint is apprecciated.
Thanks

**Comment by sazzad16** [comment]
Apologies everyone. For some reasons I was unsubscribed from this discussion.

---

@vedev9 

> Does the JedisPooled get a connection from the pool every time its methods are called and then releases it when done automatically?

Yes.

BTW, availability of a connection in the pool is controlled by ConnectionPoolConfig.

---

@dzoech 

> Do I have to initialize a connection pool as a singleton/static object and then pass this to the JedisPooled constructor?

No, you don't have to (but you can).

> What happens when I don't provide a constructor argument to JedisPooled?

Which constructor argument?

> Does it create a new connection pool for every JedisPooled instance ... ?

Yes.

> ... is there some shared pool abstracted somewhere ... ?

No.

**Reply by simon-baatz** [reply]
@sazzad16 
> > Do I have to initialize a connection pool as a singleton/static object and then pass this to the JedisPooled constructor?
> 
> No, you don't have to (but you can).

I have a somewhat related question: I found no way to use `ReliableTransaction` directly with `JedisPooled`, i.e. I think I have to create the `ConnectionProvider` separately because I did not find the equivalent to `pipelined()` in JedisPool. Am I overlooking something?

Btw. do you intend to deprecate the old JedisPool API at some point in time? If so, are there already plans when that's going to happen?

**Reply by sazzad16** [reply]
@simon-baatz 

If you already have `JedisPooled client`,
`client.getPool().getResource()` would give you a `Connection`.
You can create `ReliableTransaction` with that.
`close()` of that Connection should be called at the end to return it back to the Pool.

---

About JedisPool, there is nothing tentative.

**Reply by simon-baatz** [reply]
Thank you for the quick answer!

I see now where my problem comes from: I intend to use `JedisPooled` for local testing but actually, we use a Sentinel deployment, i.e. we need the new `JedisSentineled`  (which doesn't provide a `getPool()` although it uses a connection pool if I get this correctly).

Moreover, the common interface of the two is `UnifiedJedis` which does not provide a `getPool()`. Would it make sense to have a common interface or even base for `JedisPooled` and `JedisSentineled` that allows to get the underlying pool?

**Reply by sazzad16** [reply]
> Would it make sense to have a common interface or even base for JedisPooled and JedisSentineled that allows to get the underlying pool?

Yes.

**Comment by siranjeeviajith** [comment]
Hi @sazzad16 and @dbeken85 , does this JedisPooled is reliable for multi-threaded applications? can we use Single JedisPooled instance for all threads?

can we rely on it to handle the connections and object pool handling by itself? in case of connection loss, do we have to do anything in order to handle pools for Jedis objects and connections?

**Reply by sazzad16** [reply]
> does this JedisPooled is reliable for multi-threaded applications? can we use Single JedisPooled instance for all threads?
> 
> can we rely on it to handle the connections and object pool handling by itself? in case of connection loss, do we have to do anything in order to handle pools for Jedis objects and connections?

Yes. Yes. Yes. No; not the Jedis objects and connections.

@siranjeeviajith But you should _tune_ `GenericObjectPoolConfig<Connection>` to adapt to the load of your application.
